### PR TITLE
Add simple effect types to Jaxprs (based on #8699)

### PR DESCRIPTION
**Pull request by sharadmv**
Adds simple effect types to Jaxprs (based on #8699).

This PR is adds a new fields to `Jaxpr`s and `JaxprEqn`s, `effects`, which is a `Set[Any]`. This PR adds the relevant threading of `effects` through the Jaxpr transformations and also adds a `.replace` convenience method for `JaxprEqn`s, `Jaxpr`s and `ClosedJaxpr`s.

Effects are added to Jaxpr at trace time via the `abstract_eval` rule, whose return values are augmented from `List[Aval]` to `Tuple[List[Aval], Set[Any]]`. Rather than rewriting every `abstract_eval` rule, we have `def_abstract_eval` wrap the provided function with one that additionally returns an empty set. Primitives now have a new `def_effectful_abstract_eval` that enables providing `abstract_eval` methods that return both avals and effects.

This PR also disallows effects in control flow primitives (i.e. `cond`, `while_loop`, and `scan`) by erroring if an effect is detected after tracing the body functions to jaxprs.

### CODE REVIEW

File: jax/_src/lax/control_flow.py [hunk 1]
```diff
@@ -1385,6 +1396,8 @@ def _cond_typecheck(*avals, branches, linear):
     raise core.JaxprTypeError(
       f'cond branches take input types {jaxpr0_in_avals_str}, '
       f'called with operands of type {_avals_short(op_avals)}')
+  effects = core.join_effects((b.effects for b in branches))
```
Comment (mattjj) [inline jax/_src/lax/control_flow.py:1399]: I think we want to check that the effects are the same across branches here. That is, something like

```python
if not all(branches[0].effects == b.effects for b in branches[1:]):
  raise ...
```

File: jax/core.py [hunk 2]
```diff
@@ -91,6 +96,18 @@ def pretty_print(self, *, source_info=False, print_shapes=True,
...
+    constvars = constvars or self.constvars
+    invars = invars or self.invars
+    outvars = outvars or self.outvars
+    eqns = eqns or self.eqns
+    effects = effects or self.effects
+    return Jaxpr(constvars=constvars, invars=invars, outvars=outvars, eqns=eqns,
+                 effects=effects)
+
+def join_effects(effects: Iterable[Effects]) -> Effects:
+  return functools.reduce(lambda a, b: a | b, effects, set())
```
Comment (mattjj) [inline jax/core.py:110]:
```suggestion
  return set.union(*effects)
```

(Though TBH I kind of like not introducing a new abstraction in core.py and instead just calling `set.union` where `core.join_effects` is currently called. Up to you!)

File: jax/core.py [hunk 3]
```diff
@@ -161,17 +187,22 @@ class JaxprEqn(NamedTuple):
...
   def __repr__(self):
     return str(pp_eqn(self, JaxprPpContext(), custom_pp_eqn_rules=False)
                ).rstrip()
 
-def new_jaxpr_eqn(invars, outvars, primitive, params, source_info=None):
+  def replace(self, *args, **kwargs):
+    return self._replace(*args, **kwargs)
+
+def new_jaxpr_eqn(invars, outvars, primitive, params, effects, source_info=None):
+  effects = effects or set()
```
Comment (mattjj) [inline jax/core.py:201]: Do we want to support any falsey value to be passed to new_jaxpr_eqn? Maybe better not to add that polymorphism and require the caller to pass `set()`.

File: jax/experimental/djax.py [hunk 4]
```diff
@@ -954,9 +955,10 @@ def new_res(v):
     unks = map(read, eqn.invars)
     if any(unks):
       invars = [v if unk else new_res(v) for unk, v in zip(unks, eqn.invars)]
-      eqns2.append(pe.new_jaxpr_eqn(invars, eqn.outvars, eqn.primitive,
-                                    eqn.params,
-                                    source_info_util.new_source_info()))
+      eqns2.append(core.new_jaxpr_eqn(invars, eqn.outvars, eqn.primitive,
+                                      eqn.params,
+                                      None,
```
Comment (mattjj) [inline jax/experimental/djax.py:960]:
```suggestion
                                      set(),
```

Comment (mattjj) [reply jax/experimental/djax.py:960]: An alternative to writing `set()` literals everywhere would be to define a `no_effects` variable and use that.

Also maybe join the lines here.

File: jax/_src/lax/control_flow.py [hunk 5]
```diff
@@ -326,8 +330,9 @@ def _create_jaxpr(init_val):
                       body_nconsts=len(body_consts), body_jaxpr=body_jaxpr)
   return tree_unflatten(body_tree, outs)
 
-def _while_loop_abstract_eval(*args, **kwargs):
-  return _map(raise_to_shaped, kwargs["body_jaxpr"].out_avals)
+def _while_loop_abstract_eval(*args, body_jaxpr, **kwargs):
+  del args, kwargs
+  return _map(raise_to_shaped, body_jaxpr.out_avals), body_jaxpr.effects
```
Comment (mattjj) [inline jax/_src/lax/control_flow.py:335]: Can the cond_jaxpr have effects? The code above suggests yes. Then the effect of applying a while_loop isn't just the effect of the body.

I think for this PR we should just raise NotImplementedError in the `while_loop` traceable if the jaxprs formed have any effects at all. Then let's add support for effects to each control flow primitive one by one.

File: jax/_src/lax/control_flow.py [hunk 6]
```diff
@@ -315,6 +316,9 @@ def _create_jaxpr(init_val):
     new_init_val, = tree_unflatten(in_tree, new_init_vals)
     init_vals, init_avals, body_jaxpr, in_tree, *rest = _create_jaxpr(new_init_val)
   cond_jaxpr, cond_consts, body_consts, body_tree = rest
+  joined_effects = core.join_effects([body_jaxpr.effects, cond_jaxpr.effects])
+  body_jaxpr = body_jaxpr.replace(jaxpr=body_jaxpr.jaxpr.replace(effects=joined_effects))
+  cond_jaxpr = cond_jaxpr.replace(jaxpr=cond_jaxpr.jaxpr.replace(effects=joined_effects))
```
Comment (mattjj) [inline jax/_src/lax/control_flow.py:321]: Also mentioned below: let's just raise a NotImplementedError if any control flow primitives get jaxprs with effects in them (in the traceables, maybe also in the typecheck rules). Getting these right will take some care so we should probably add support for control flow primitives one-by-one in follow-up PRs.

To put a fine point on it: I think the only changes this PR should make to control_flow.py should be to raise NotImplementedErrors (if this PR makes any changes at all).

File: jax/_src/lax/lax.py [hunk 7]
```diff
@@ -2715,7 +2715,7 @@ def _broadcast_in_dim_staging_rule(
   out_tracer = pe.DynamicJaxprTracer(trace, aval, source_info)
   invars = [trace.getvar(x), *(trace.getvar(d) for d in dyn_shape)]
   eqn = pe.new_jaxpr_eqn(invars, [trace.makevar(out_tracer)],
-                         broadcast_in_dim_p, params, source_info)
+                         broadcast_in_dim_p, params, None, source_info)
```
Comment (mattjj) [inline jax/_src/lax/lax.py:2718]:
```suggestion
                         broadcast_in_dim_p, params, set(), source_info)
```

File: jax/experimental/djax.py [hunk 8]
```diff
@@ -1326,8 +1328,8 @@ def _iota_staging_rule(trace, tracers, params):
     out_aval = core.ShapedArray((n,), np.dtype('int32'))
     out_tracer = pe.DynamicJaxprTracer(trace, out_aval, None)
     outvar = trace.makevar(out_tracer)
-    eqn = pe.new_jaxpr_eqn([], [outvar], iota_p, dict(size=n),
-        source_info_util.new_source_info())
+    eqn = core.new_jaxpr_eqn([], [outvar], iota_p, dict(size=n),
+        None, source_info_util.new_source_info())
```
Comment (mattjj) [inline jax/experimental/djax.py:1332]:
```suggestion
        set(), source_info_util.new_source_info())
```

File: jax/experimental/djax.py [hunk 9]
```diff
@@ -1340,8 +1342,8 @@ def _iota_staging_rule(trace, tracers, params):
     out_tracer = pe.DynamicJaxprTracer(trace, out_aval, None)
     outvar = trace.makevar(out_tracer)
     invar = trace.getvar(tracer)
-    eqn = pe.new_jaxpr_eqn([invar], [outvar], iota_p, {},
-        source_info_util.new_source_info())
+    eqn = core.new_jaxpr_eqn([invar], [outvar], iota_p, {},
+        None, source_info_util.new_source_info())
```
Comment (mattjj) [inline jax/experimental/djax.py:1346]:
```suggestion
        set(), source_info_util.new_source_info())
```

File: jax/experimental/djax.py [hunk 10]
```diff
@@ -1389,9 +1391,10 @@ def _broadcast_staging_rule(trace, tracers, params):
     dtype = aval._eltTy._dtype if isinstance(aval, AbsArray) else aval.dtype
     out_aval = AbsArray((d, *x.shape), BaseType(dtype))
     out_tracer = pe.DynamicJaxprTracer(trace, out_aval, None)
-    eqn = pe.new_jaxpr_eqn([trace.getvar(x), trace.getvar(d)],
-                           [trace.makevar(out_tracer)], broadcast_p, {},
-                           source_info_util.new_source_info())
+    eqn = core.new_jaxpr_eqn([trace.getvar(x), trace.getvar(d)],
+                             [trace.makevar(out_tracer)], broadcast_p, {},
+                             None,
```
Comment (mattjj) [inline jax/experimental/djax.py:1396]:
```suggestion
                             set(),
```

File: jax/experimental/host_callback.py [hunk 11]
```diff
@@ -1420,10 +1420,10 @@ def _rewrite_jaxpr(jaxpr: core.Jaxpr, has_input_token: bool,
     # We need tokens but none is given in input; make one depending on all invars
     eqns.append(
         core.new_jaxpr_eqn(jaxpr.invars, [last_token_var],
-                           lax.create_token_p, {}, source_info_util.current()))
+                           lax.create_token_p, {}, None, source_info_util.current()))
```
Comment (mattjj) [inline jax/experimental/host_callback.py:1423]:
```suggestion
                           lax.create_token_p, {}, set(), source_info_util.current()))
```

File: jax/experimental/host_callback.py [hunk 12]
```diff
@@ -1420,10 +1420,10 @@ def _rewrite_jaxpr(jaxpr: core.Jaxpr, has_input_token: bool,
...
     eqns.append(
         core.new_jaxpr_eqn(jaxpr.invars, [last_itoken_var],
-                           lax.create_token_p, {}, source_info_util.current()))
+                           lax.create_token_p, {}, None, source_info_util.current()))
```
Comment (mattjj) [inline jax/experimental/host_callback.py:1426]:
```suggestion
                           lax.create_token_p, {}, set(), source_info_util.current()))
```

**Review by mattjj — APPROVED** [review]
LGTM, my main suggestion is not to touch lax_control_flow.py in this PR (or just add NotImplementedErrors to it).

The eqn.replace / jaxpr.replace stuff is a great QOL improvement!

File: jax/_src/lax/control_flow.py (hunk 1 above)
Comment (sharadmv) [reply jax/_src/lax/control_flow.py]: Done.

File: jax/core.py (hunk 2 above)
Comment (sharadmv) [reply jax/core.py]: Done.

File: jax/experimental/djax.py (hunk 4 above)
Comment (sharadmv) [reply jax/experimental/djax.py]: Done.

File: jax/_src/lax/control_flow.py (hunk 5 above)
Comment (sharadmv) [reply jax/_src/lax/control_flow.py]: Done.

File: jax/_src/lax/lax.py (hunk 7 above)
Comment (sharadmv) [reply jax/_src/lax/lax.py]: Done.

File: jax/_src/lax/control_flow.py (hunk 6 above)
Comment (sharadmv) [reply jax/_src/lax/control_flow.py]: Done.

File: jax/core.py (hunk 3 above)
Comment (sharadmv) [reply jax/core.py]: Done.

File: tests/jaxpr_effects_test.py [hunk 13]
```diff
@@ -0,0 +1,133 @@
+# Copyright 2021 Google LLC
```
Comment (mattjj) [inline tests/jaxpr_effects_test.py:1]:
```suggestion
# Copyright 2022 Google LLC
```

**Review by mattjj — APPROVED** [review]

File: jax/experimental/host_callback.py [hunk 14]
```diff
@@ -1476,11 +1476,9 @@ def _rewrite_eqn(eqn: core.JaxprEqn, eqns: List[core.JaxprEqn],
   """
```
Comment (gnecula) [inline jax/experimental/host_callback.py:1476]: This whole code ought to disappear once we have support for effects in Jaxpr.

Comment (sharadmv) [reply jax/experimental/host_callback.py:1476]: Yes, we should be able to rewrite HCB once we can lower effects away with tokens.

File: jax/core.py [hunk 15]
```diff
@@ -54,14 +54,21 @@
...
+Effects = Set[Effect]
+
+no_effects: Effects = set()
+
 class Jaxpr:
   constvars: List[Var]
   invars: List[Var]
   outvars: List[Atom]
   eqns: List[JaxprEqn]
+  effects: Effects
```
[the remark this answers is not in the document]
Comment (sharadmv) [reply jax/core.py:67]: I've added some more detail to the constructor docstring.

File: jax/experimental/host_callback.py (hunk 14 above)
Comment (sharadmv) [reply jax/experimental/host_callback.py:1476]: Yes, that's the goal

**Pull request merged**